index.js

```diff
@@ -1,186 +1,259 @@
-    process.once('exit', function() {
-      simProcess.kill('SIGTERM');
-    });
+        var simulator = results[0];
+        port = results[1];
+
+        launchSimulatorAndWaitUntilReady({
+          binary: simulator.bin,
+          profile: simulator.profile,
+          port: port,
+          detached: detached,
+          verbose: verbose
+        }).then(function(simulatorDetails) {
+          resolve(simulatorDetails);
+        }, function(simulatorLaunchError) {
+          reject(simulatorLaunchError);
+        });
+
+      }, function(error) {
+        reject(error);
+      });
 
-    process.once('uncaughtException', function(error) {
-      if (process.listeners('uncaughtException').length === 0) {
-        simProcess.kill('SIGTERM');
-        throw error;
+  });
+}
+
+// Find a simulator that matches the options
+function findSimulator(options) {
+
+  return new Promise(function(resolve, reject) {
+
+    findSimulators(options).then(function(results) {
+
+      if (!results || results.length === 0) {
+        reject(new Error('No simulators installed, or cannot find them'));
       }
+
+      // just returning the first result for now
+      resolve(results[0]);
+
+    }, function(error) {
+      reject(error);
     });
-  }
 
-  if (opts.exit) {
-    simProcess.unref();
-  }
-  defer.resolve(simProcess);
-  return defer.promise;
+  });
+
 }
 
-function createClient(simulator) {
-  var deferred = Q.defer();
-  var client = new FirefoxClient();
-  client.connect(simulator.port, function(err) {
-    if (err) {
-      deferred.reject(err);
+
+function findAvailablePort(preferredPort) {
+
+  return new Promise(function(resolve, reject) {
+
+    // Start searching with the preferred port, if specified
+    if (preferredPort !== undefined) {
+      portFinder.basePort = preferredPort;
     }
-    simulator.client = client;
-    deferred.resolve(simulator);
+    
+    portFinder.getPort(function(err, port) {
+      if (err) {
+        reject(err);
+      } else {
+        console.log('got this port', port);
+        resolve(port);
+      }
+    });
   });
-  return deferred.promise;
+
 }
 
-function runB2G(opts) {
-  var commandReady = commandB2G(opts);
-  var portReady = commandReady.then(portIsReady.bind(null, opts.port));
-  return portReady.then(function() {
-    return commandReady;
+
+// Launches the simulator and wait until it's ready to be used
+function launchSimulatorAndWaitUntilReady(options) {
+
+  var port = options.port;
+  var binary = options.binary;
+  var profile = options.profile;
+
+  return new Promise(function(resolve, reject) {
+
+    launchSimulator(options).then(function(simulatorProcess) {
+      waitUntilSimulatorIsReady(port).then(function() {
+        resolve({
+          process: simulatorProcess,
+          pid: simulatorProcess.pid,
+          port: port,
+          binary: binary,
+          profile: profile
+        });
+      }, function(timedOutError) {
+          reject(timedOutError);
```

Comment:
The `reject(timedOutError);` is four spaces instead of two. It reads:

```javascript
}, function(timedOutError) {
    reject(timedOutError);
});
```
but should be:

```javascript
}, function(timedOutError) {
  reject(timedOutError);
});
```